src/main/kotlin/org/rust/cargo/runconfig/buildtool/CargoBuildListener.kt

```diff
@@ -0,0 +1,109 @@
+import com.intellij.openapi.util.Key
+import com.intellij.openapi.vfs.VfsUtil
+import org.rust.cargo.CargoConstants
+import org.rust.cargo.runconfig.RsAnsiEscapeDecoder
+import org.rust.cargo.runconfig.RsExecutableRunner.Companion.binaries
+import org.rust.cargo.runconfig.createFilters
+
+@Suppress("UnstableApiUsage")
+class CargoBuildListener(
+    private val context: CargoBuildContext,
+    private val buildProgressListener: BuildProgressListener
+) : ProcessAdapter(), AnsiEscapeDecoder.ColoredTextAcceptor {
+    private val decoder: AnsiEscapeDecoder = RsAnsiEscapeDecoder()
+
+    private val buildOutputParser: CargoBuildEventsConverter = CargoBuildEventsConverter(context)
+
+    private val instantReader = BuildOutputInstantReaderImpl(
+        context.buildId,
+        context.buildId,
+        buildProgressListener,
+        listOf(buildOutputParser)
+    )
+
+    private val lineBuffer: MutableList<String> = mutableListOf()
+
+    override fun startNotified(event: ProcessEvent) {
+        val descriptor = DefaultBuildDescriptor(
+            context.buildId,
+            "Run Cargo command",
+            context.workingDirectory.toString(),
+            context.started
+        )
+        val buildStarted = StartBuildEventImpl(descriptor, "${context.taskName} running...")
+            .withExecutionFilters(*createFilters(context.cargoProject).toTypedArray())
+            .withRestartAction(createStopAction(event.processHandler))
+        buildProgressListener.onEvent(context.buildId, buildStarted)
+    }
+
+    override fun processTerminated(event: ProcessEvent) {
+        instantReader.closeAndGetFuture().whenComplete { _, _ ->
+            val succeeded = event.exitCode == 0
+            context.environment.binaries = buildOutputParser.binaries.takeIf { succeeded }
+            context.finished(succeeded)
+
+            val result = context.result.get()
+            val buildFinished = FinishBuildEventImpl(
+                context.buildId,
+                null,
+                context.finished,
+                result.message,
+                if (result.succeeded) SuccessResultImpl() else FailureResultImpl(null as Throwable?)
+            )
+            buildProgressListener.onEvent(context.buildId, buildFinished)
+
+            val targetPath = context.workingDirectory.resolve(CargoConstants.ProjectLayout.target)
+            val targetDir = VfsUtil.findFile(targetPath, true) ?: return@whenComplete
+            VfsUtil.markDirtyAndRefresh(true, true, true, targetDir)
+        }
+    }
+
+    override fun onTextAvailable(event: ProcessEvent, outputType: Key<*>) {
+        lineBuffer.add(event.text)
+        if (event.text.endsWith('\n')) {
+            // progress message [K actual message
+            val lines = lineBuffer.joinToString("").split("[K")
+            lineBuffer.clear()
+
+            // don't print progress message
+            buildOutputParser.parseOutput(lines.last(), outputType == ProcessOutputTypes.STDOUT) {
```

Comment:
Can `lines` be empty (`NoSuchElementException`)?

Reply:
No, see line 80.